pkg/virt-controller/watch/vmi.go

```diff
@@ -1039,3 +1114,303 @@ func (c *VMIController) currentPod(vmi *virtv1.VirtualMachineInstance) (*k8sv1.P
+	if len(deletedVolumes) > 0 {
+		// Some volumes were deleted, make sure we delete the hotplug pods
+		for _, volume := range deletedVolumes {
+			logger.V(1).Infof("Deleting attachment pod for volume: %s", volume.Name)
+			deletedPod, err := c.deleteAttachmentPodForVolume(vmi, volume, hotplugAttachmentPods)
+			if err != nil {
+				return nil, &syncErrorImpl{fmt.Errorf("Error deleting pod %s/%s, %v", deletedPod.Namespace, deletedPod.Name, err), FailedDeletePodReason}
+			}
+			if deletedPod != nil {
+				c.recorder.Eventf(vmi, k8sv1.EventTypeNormal, SuccessfulDeletePodReason, "Deleted hotplug attachment pod %s/%s, for volume %s", deletedPod.Namespace, deletedPod.Name, volume.Name)
+			}
+			// Not deleting volume status, updating it to indicate the pod was deleted, once the pod is really gone, the volume status will be
+			// removed.
+			for _, statusVolume := range statusCopy {
+				if statusVolume.Name == volume.Name && statusVolume.HotplugVolume != nil {
+					statusVolume.HotplugVolume.Phase = virtv1.HotplugVolumeDetaching
+					statusVolume.HotplugVolume.Reason = SuccessfulDeletePodReason
+					statusVolume.HotplugVolume.Message = fmt.Sprintf("Deleted hotplug attachment pod %s/%s, for volume %s", deletedPod.Namespace, deletedPod.Name, volume.Name)
+				}
+			}
+		}
+	}
+	newVolumes := c.getNewHotplugVolumes(hotplugAttachmentPods, hotplugVolumes)
+	if len(newVolumes) > 0 {
+		// New volumes detected, create hotplug pods.
+		for _, volume := range newVolumes {
+			logger.V(1).Infof("Processing new hotplugged volume: %s", volume.Name)
+			var hotplugPodTemplate *k8sv1.Pod
+			var err error
+			// Check if the VMI VolumeStatus contains this volume, if that is the case then something deleted the attachment pod
+			// and we need to stop the VMI as that is a critical error.
+			if c.volumeStatusContainsVolumeAndPod(vmi.Status.VolumeStatus, volume) {
+				logger.V(1).Infof("Detected attachment pod is missing for VMI %s/%s, the VMI will be deleted", vmi.Namespace, vmi.Name)
+				return nil, &syncErrorImpl{fmt.Errorf("Missing pod for hotplugged volume %s", volume.Name), MissingAttachmentPodReason}
+			}
+			hotplugPodTemplate, statusCopy, err = c.createAttachmentPodTemplate(volume, virtLauncherPod, vmi, statusCopy)
+			if err != nil {
+				return nil, &syncErrorImpl{fmt.Errorf("Error creating attachment pod template %v", err), FailedCreatePodReason}
+			}
+			if hotplugPodTemplate != nil { // nil means the PVC is not populated yet.
+				pod, err := c.clientset.CoreV1().Pods(vmi.GetNamespace()).Create(hotplugPodTemplate)
+				if err != nil && !k8serrors.IsAlreadyExists(err) {
+					c.recorder.Eventf(vmi, k8sv1.EventTypeWarning, FailedCreatePodReason, "Error creating hotplug pod for volume %s: %v", volume.Name, err)
+					return nil, &syncErrorImpl{fmt.Errorf("Error creating attachment pod %s/%s, %v", hotplugPodTemplate.Namespace, hotplugPodTemplate.Name, err), FailedCreatePodReason}
+				}
```

Comment:
We'll need to use expectations here. take a look at our usage of the ```c.podExpectations.ExpectCreations(vmiKey, 1)``` and ``` c.podExpectations.ExpectDeletions(vmiKey, []string{controller.PodKey(pod)})``` functions.

What the expectation logic does is ensure that our `sync()` logic does not execute until our informers have observed the actions performed by the previous sycn() execution... So if we add a pod, we set an expectation that a new pod should be observed, and we won't execute sync() again until that pod is observed. Otherwise, we could end up in scenarios where we accidentally create multiple pods because we didn't detect that a previous pod already exists in the informer

The IsAlreadyExists() check doesn't help here because you're using generateName on the pod. We'll likely end up with multiple pods in certain scenarios without the expectation.